chore(share|das|libs/utils): DRY context resetting in the event of an error for metrics

cyneptic opened the pull request:
this creates a function in the utils/resetctx.go that takes a context as an input and resets it if there is an error.
it then uses the function to refactor all the instances of the repeating pattern as requested in the issue - fixes #2928

renaynay commented on libs/utils/resetctx.go:
```diff
@@ -0,0 +1,12 @@
+
+import "context"
+
+// ResetContextOnError returns the context if there is no error, otherwise returns it after resetting it to background.
```
```suggestion
// ResetContextOnError returns a fresh context if the given context has an error. 
```

renaynay reviewed (commented):
This is fine with me :) thanks @cyneptic 

Do you mind rebasing on main once https://github.com/celestiaorg/celestia-node/pull/2863

is merged and applying the same to the core package?

cyneptic commented:
> This is fine with me :) thanks @cyneptic
> 
> Do you mind rebasing on main once #2863
> 
> is merged and applying the same to the core package?

Hi, thank you. tried rebasing and was already up-to-date.
as for the core package, I didn't find any instances of the pattern in there. I applied this to all the patterns I found in the project.

cyneptic commented:
> This is fine with me :) thanks @cyneptic
> 
> Do you mind rebasing on main once #2863
> 
> is merged and applying the same to the core package?

Hello and thank you.

rebased, is merged to main and applied to core.

also added the suggested change to the function comment.

thank you :)

renaynay reviewed (approved):
lgtm, thanks @cyneptic !

renaynay reviewed (commented):
@cyneptic Can you please lint 🙏🏻

cyneptic commented:
> @cyneptic Can you please lint 🙏🏻

hey, i fixed the 6 lint issues that didn't pass last time. golanci-lint doesn't return any errors now, hopefully its okay.

btw github says "dismissed renaynay's stale review" im not sure how that happened or what it is to be honest 😄

cyneptic commented:
so, whats the status on this pr ? do you want me to update it?

renaynay reviewed (approved):
All good with me :)

ramin commented:
closes #2928
